src/kernels/workaround_issue_1431.hpp

```diff
@@ -0,0 +1,37 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2022 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#ifndef WORKAROUND_ISSUE_1431_HPP
+#define WORKAROUND_ISSUE_1431_HPP
+
+// This is for the case if compiler defines `warpSize` as a macro.
+#ifdef warpSize
+#undef warpSize
+#endif
+
+// Hack: this replaces each `warpSize` by literal `32`.
+#define warpSize 32
```

Comment:
@atamazov @JehandadKhan  it seems an order issue, still conflicting with `hip_runtime.h`

```
[2022-04-12T02:44:44.578Z] ../bin/test_reduce_test --half --D 7 1024 1 --R 0 1 2 --ReduceOp 0 --CompType 1 --N 0 --I 0 --scales 1 0 
[2022-04-12T02:44:44.578Z] FAILED: /var/jenkins/workspace/MLLibs_MIOpen_PR-1469/src/hipoc/hipoc_program.cpp:300: Code object build failed. Source: gridwise_generic_reduction_second_call_warpwise_reduce_all_dims.cpp
[2022-04-12T02:44:44.578Z] verify_reduce_no_indices failed
[2022-04-12T02:44:44.578Z] Input Tensor 7, 1024, 1
[2022-04-12T02:44:44.578Z] MIOpen(HIP): Error [Do] 'amd_comgr_do_action(kind, handle, in.GetHandle(), out.GetHandle())' AMD_COMGR_ACTION_COMPILE_SOURCE_TO_BC: ERROR (1)
[2022-04-12T02:44:44.578Z] MIOpen(HIP): Error [BuildHip] comgr status = ERROR (1)
[2022-04-12T02:44:44.578Z] MIOpen(HIP): Warning [BuildHip] In file included from /tmp/comgr-843102/input/gridwise_generic_reduction_second_call_warpwise_reduce_all_dims.cpp:29:
[2022-04-12T02:44:44.578Z] In file included from /tmp/comgr-843102/include/config.hpp:5:
[2022-04-12T02:44:44.578Z] In file included from /opt/rocm-4.3.1/hip/../include/hip/hip_runtime.h:62:
[2022-04-12T02:44:44.578Z] In file included from /opt/rocm-4.3.1/hip/../include/hip/amd_detail/hip_runtime.h:63:
[2022-04-12T02:44:44.578Z] /opt/rocm-4.3.1/hip/../include/hip/hip_runtime_api.h:89:9: error: expected member name or ';' after declaration specifiers
[2022-04-12T02:44:44.578Z]     int warpSize;              ///< Warp size.
[2022-04-12T02:44:44.578Z]     ~~~ ^
[2022-04-12T02:44:44.578Z] /tmp/comgr-843102/include/workaround_issue_1431.hpp:35:18: note: expanded from macro 'warpSize'
[2022-04-12T02:44:44.578Z] #define warpSize 32
[2022-04-12T02:44:44.578Z]                  ^
[2022-04-12T02:44:44.578Z] In file included from /tmp/comgr-843102/input/gridwise_generic_reduction_second_call_warpwise_reduce_all_dims.cpp:29:
[2022-04-12T02:44:44.578Z] In file included from /tmp/comgr-843102/include/config.hpp:5:
[2022-04-12T02:44:44.578Z] In file included from /opt/rocm-4.3.1/hip/../include/hip/hip_runtime.h:62:
[2022-04-12T02:44:44.578Z] In file included from /opt/rocm-4.3.1/hip/../include/hip/amd_detail/hip_runtime.h:70:
[2022-04-12T02:44:44.578Z] In file included from /opt/rocm-4.3.1/hip/../include/hip/amd_detail/hip_atomic.h:3:
[2022-04-12T02:44:44.578Z] /opt/rocm-4.3.1/hip/../include/hip/amd_detail/device_functions.h:295:22: error: expected unqualified-id
[2022-04-12T02:44:44.578Z] static constexpr int warpSize = __AMDGCN_WAVEFRONT_SIZE;
[2022-04-12T02:44:44.578Z]                      ^
[2022-04-12T02:44:44.578Z] /tmp/comgr-843102/include/workaround_issue_1431.hpp:35:18: note: expanded from macro 'warpSize'
[2022-04-12T02:44:44.578Z] #define warpSize 32
[2022-04-12T02:44:44.578Z]                  ^
[2022-04-12T02:44:44.578Z] 2 errors generated when compiling for gfx1030.

Reply:
@atamazov It seem to be the same issues discussed at https://github.com/ROCmSoftwarePlatform/MIOpen/issues/1431

Reply:
The W/A in this PR is working incorrectly when `MIOPEN_DEBUG_COMGR_HIP_PCH_ENFORCE=0`.

The reason is a coding bug in my PR. W/A for issue #1431 should not be enabled when PCH is not enabled. Currently W/A is not enabled when _"PCH is not supported"_ but this is not the same as _"PCH is not enabled"_. For example, PCH is supported but not enabled when `MIOPEN_DEBUG_COMGR_HIP_PCH_ENFORCE=0` in the environment.

It seems like the failing test is running with `MIOPEN_DEBUG_COMGR_HIP_PCH_ENFORCE=0` (it comes from https://github.com/atamazov/MIOpen/blob/c330fd9ec8c4cdf77783ffd5f96d08bf37f34d9b/test/CMakeLists.txt#L277 from my fork, which is outdated and still contains W/A for issue #898 from where this env var came from).

I'll fix the bug and update PR from the recent develop.

Reply:
#1508 is already merged so maybe update your branch with the latest develop?

Reply:
Fix is also required. Please expect updated PR tomorrow.

Reply:
@atamazov  last normal promotion should be started today so I think this one would likely to be a cherry picking candidate instead. :)

Reply:
Understood, thanks. BTW fix is in developement/testing still...

Reply:
Resolved in 65ca9f2229dd6885cc83eb316f25cc2d88fa70a7